components/brave_wallet/browser/internal/hd_key_ed25519.cc

```diff
@@ -62,74 +74,74 @@ std::string HDKeyEd25519::GetPath() const {
+  if (index >= kHardenedOffset) {
     return nullptr;
   }
-  auto child_path =
-      base::StrCat({path_, "/", base::NumberToString(index), "'"});
-  return std::make_unique<HDKeyEd25519>(std::move(child_path),
-                                        std::move(child_private_key));
+  return DeriveChild(kHardenedOffset + index);
+}
+
+std::unique_ptr<HDKeyEd25519> HDKeyEd25519::DeriveChild(uint32_t index) {
+  std::vector<uint8_t> hmac_payload(37);
+
+  auto span_writer = base::SpanWriter(base::span(hmac_payload));
+  // https://github.com/satoshilabs/slips/blob/master/slip-0010.md#private-parent-key--private-child-key
+  span_writer.Write(base::U8ToBigEndian(0));
+  span_writer.Write(GetPrivateKeyAsSpan());
+  span_writer.Write(base::U32ToBigEndian(index));
+
+  return base::WrapUnique(new HDKeyEd25519(chain_code_, hmac_payload));
 }
 
-std::unique_ptr<HDKeyEd25519> HDKeyEd25519::DeriveChildFromPath(
-    const std::string& path) {
-  if (path_ != kMasterNode) {
-    VLOG(0) << "must derive only from master key";
+// static
+std::unique_ptr<HDKeyEd25519> HDKeyEd25519::GenerateFromSeedAndPath(
+    base::span<const uint8_t> seed,
+    std::string_view hd_path) {
+  auto hd_key = base::WrapUnique(
+      new HDKeyEd25519(base::byte_span_from_cstring(kMasterSecret), seed));
+
+  auto nodes = ParseFullHDPath(hd_path);
+  if (!nodes) {
     return nullptr;
   }
 
-  auto child_private_key = private_key_->unwrap().derive(path);
-  if (!child_private_key->is_ok()) {
-    VLOG(0) << std::string(child_private_key->error_message());
-    return nullptr;
+  for (auto index : *nodes) {
+    if (index < kHardenedOffset) {
+      return nullptr;
+    }
+    hd_key = hd_key->DeriveChild(index);
+    if (!hd_key) {
+      return nullptr;
+    }
   }
 
-  return std::make_unique<HDKeyEd25519>(path, std::move(child_private_key));
+  return hd_key;
 }
 
-std::vector<uint8_t> HDKeyEd25519::Sign(base::span<const uint8_t> msg) {
-  auto signature_result = private_key_->unwrap().sign(
-      rust::Slice<const uint8_t>{msg.data(), msg.size()});
-  if (!signature_result->is_ok()) {
-    VLOG(0) << std::string(signature_result->error_message());
-    return std::vector<uint8_t>();
-  }
-  auto signature_bytes = signature_result->unwrap().to_bytes();
-  return std::vector<uint8_t>(signature_bytes.begin(), signature_bytes.end());
-}
+std::array<uint8_t, kEd25519SignatureSize> HDKeyEd25519::Sign(
+    base::span<const uint8_t> msg) {
+  std::array<uint8_t, kEd25519SignatureSize> signature = {};
 
-bool HDKeyEd25519::VerifyForTesting(base::span<const uint8_t> msg,
-                                    base::span<const uint8_t> sig) {
-  auto verification_result = private_key_->unwrap().verify(
-      rust::Slice<const uint8_t>{msg.data(), msg.size()},
-      rust::Slice<const uint8_t>{sig.data(), sig.size()});
-  if (!verification_result->is_ok()) {
-    VLOG(0) << std::string(verification_result->error_message());
-    return false;
-  }
-  return true;
+  CHECK(
```

Comment:
Do we want a browser crash here if this doesn't sign properly? I don't see a reason that this would occur, but I'm thinking it would be safer to not crash completely if say the message isn't as expected or something like that.

Reply:
ok, will fix